contracts/airdrops.js

```diff
@@ -0,0 +1,244 @@
+      params.maxTransactionsPerBlock = maxTransactionsPerBlock;
+    }
+    if (maxAirdropsPerBlock) {
+      if (!api.assert(Number.isInteger(maxAirdropsPerBlock) && maxAirdropsPerBlock >= 1, 'invalid maxAirdropsPerBlock')) return;
+      params.maxAirdropsPerBlock = maxAirdropsPerBlock;
+    }
+
+    await api.db.update('params', params);
+  }
+};
+
+const parseAirdrop = async (list, precision) => {
+  const params = await api.db.findOne('params', {});
+  const airdrop = {};
+  airdrop.list = [];
+  airdrop.fee = '0';
+  airdrop.quantity = '0';
+  airdrop.isValid = false;
+
+  // loop through list and validate
+  for (let i = 0; i < list.length; i += 1) {
+    const { 0: to, 1: quantity } = list[i];
+
+    if (to && api.isValidAccountName(to)
+      && quantity && !api.BigNumber(quantity).isNaN()
+      && api.BigNumber(quantity).gt(0) && api.BigNumber(quantity).dp() <= precision) {
+      airdrop.list.push({
+        to,
+        quantity,
+      });
+
+      // add this quantity to the total quantity of tokens to airdrop
+      airdrop.quantity = api.BigNumber(airdrop.quantity).plus(quantity);
+    }
+  }
+
+  // calculate total fee
+  airdrop.fee = api.BigNumber(params.feePerTransaction).times(airdrop.list.length);
+
+  // list validation, check if all values from list are valid & pushed into airdrop.list
+  if (list.length > 0 && list.length === airdrop.list.length) {
+    airdrop.isValid = true;
+  }
+
+  return airdrop;
+};
+
+const hasValidType = (token, type) => {
+  if (type === 'transfer') {
+    return true;
+  }
+
+  // check if staking is enabled
+  if (type === 'stake' && api.assert(token.stakingEnabled === true, 'staking not enabled')) {
+    return true;
+  }
+
+  return false;
+};
+
+const transferIsSuccesfull = (result, action, from, to, symbol, quantity) => {
+  if (result.errors === undefined
+    && result.events && result.events.find(el => el.contract === 'tokens'
+    && el.event === action
+    && el.data.from === from
+    && el.data.to === to
+    && api.BigNumber(el.data.quantity).eq(quantity)
+    && el.data.symbol === symbol) !== undefined) {
+    return true;
+  }
+
+  return false;
+};
+
+actions.newAirdrop = async (payload) => {
+  const {
+    symbol,
+    type,
+    list,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (api.assert(isSignedWithActiveKey === true, 'you must use a custom_json signed with your active key')
+    && api.assert(symbol && typeof symbol === 'string'
+      && type && typeof type === 'string'
+      && list && typeof list === 'object' && list.length, 'invalid params')
+    && api.assert(type === 'transfer' || type === 'stake', 'invalid type')) {
+    const token = await api.db.findOneInTable('tokens', 'tokens', { symbol });
+
+    // get api.sender's utility and airdrop token balances
+    const { balance: utilityTokenBalance } = { ...await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: UTILITY_TOKEN_SYMBOL }) };
```

Comment:
Clearer if this is unwrapped

const utilityTokenBalance = (await api.db.findOne... ).balance;

But you probably want to store the await in some temporary variable and check if it is present. (can combine below with e.g.

const utilityTokenBalance = await api.db.findOne...
if (...  utilityTokenBalance && utilityTokenBalance.balance && api.BigNumber(utilitytokenBalance.balance).... )